File: docgen/src/lib/docGenerator.js
import { detectLanguage } from './nlpEngine'; // Assuming you have a JS equivalent of nlp_engine
import { extractRelevantSnippets } from './codeExtractor'; // Assuming you have the JS code extractor

// Function to generate markdown content in the browser
async function generateMarkdown(enhancedDocs) {
  const generatedFiles = new Set();
  
  for (const doc of enhancedDocs) {
    const filePath = doc.file_path;
    const fileName = filePath.split('/').pop();
    
    // Create a unique identifier for each file (using crypto API for hashing)
    const fileHash = await crypto.subtle.digest('MD5', new TextEncoder().encode(filePath));
    const hashArray = Array.from(new Uint8Array(fileHash));
    const hashHex = hashArray.map(b => b.toString(16).padStart(2, '0')).join('').slice(0, 8);
    const uniqueFileName = `${fileName}_${hashHex}.md`;
    
    // Construct the markdown content
    let markdownContent = `# ${fileName}\n\n`;
    markdownContent += `File path: ${filePath}\n\n`;
    markdownContent += `## Explanation\n${doc.explanation}\n\n`;
    markdownContent += `## Code Snippets\n`;

    const language = detectLanguage(filePath, doc.content);
    const snippets = extractRelevantSnippets(doc.content, language);

    snippets.forEach((snippet, index) => {
      markdownContent += `### Snippet ${index + 1}\n`;
      markdownContent += `\`\`\`${language.toLowerCase()}\n${snippet}\n\`\`\`\n\n`;
    });

    // Provide the markdown content for download or display
    const blob = new Blob([markdownContent], { type: 'text/markdown' });
    const url = URL.createObjectURL(blob);
    
    // Add the generated file to the set of downloadable files
    generatedFiles.add({
      fileName: uniqueFileName,
      url: url
    });
    
    console.log(`Generated: ${uniqueFileName}`);
  }
  
  console.log(`Generated ${generatedFiles.size} markdown files.`);
  return generatedFiles;
}

// Example usage (You could call this function when users upload or link files):
// const enhancedDocs = [
//   { file_path: "path/to/file.js", content: "file content", explanation: "This is a JS file..." }
// ];
// generateMarkdown(enhancedDocs).then(files => {
//   files.forEach(file => {
//     // You can create a download link in the DOM using file.url and file.fileName
//     console.log(`Markdown file generated: ${file.fileName} - Download URL: ${file.url}`);
//   });
// });

File: doc_generator.py
import os
import hashlib
from nlp_engine import detect_language
from code_extractor import extract_relevant_snippets

def generate_markdown(enhanced_docs, output_dir):
    generated_files = set()
    for doc in enhanced_docs:
        file_path = doc['file_path']
        file_name = os.path.basename(file_path)
        
        # Create a unique identifier for each file
        file_hash = hashlib.md5(file_path.encode()).hexdigest()[:8]
        unique_file_name = f"{file_name}_{file_hash}.md"
        
        output_path = os.path.join(output_dir, unique_file_name)
        
        try:
            with open(output_path, 'w') as f:
                f.write(f"# {file_name}\n\n")
                f.write(f"File path: {file_path}\n\n")
                f.write("## Explanation\n")
                f.write(doc['explanation'])
                f.write("\n\n")
                f.write("## Code Snippets\n")
                language = detect_language(file_path, doc['content'])
                snippets = extract_relevant_snippets(doc['content'], language)
                for i, snippet in enumerate(snippets, 1):
                    f.write(f"### Snippet {i}\n")
                    f.write(f"```{language.lower()}\n")
                    f.write(snippet)
                    f.write("\n```\n\n")
            generated_files.add(unique_file_name)
            print(f"Generated: {unique_file_name}")
        except Exception as e:
            print(f"Error generating markdown for {file_path}: {str(e)}")

    print(f"Generated {len(generated_files)} markdown files in {output_dir}")
    return generated_files
